contributors/design-proposals/dynamic-admission-control-configuration.md

```diff
@@ -0,0 +1,202 @@
+    Resources []string
+    // Subresources are the list of subresources this hook should be invoked on. '*' is all resources.
+    Subresources []string
+
+    // ClientConfig defines how to talk to the hook.
+    ClientConfig AdmissionHookClientConfig
+
+    // FailurePolicy defines how unrecognized errors from the admission endpoint are handled -
+    // allowed values are Ignore, Retry, Fail. Default value is Fail
+    FailurePolicy FailurePolicyType
+}
+
+// AdmissionHookClientConfig contains the information to make a TLS
+// connection with the webhook
+// **very similar to the schema of kubeconfig**
+type AdmissionHookClientConfig struct {
+	// Address of the external admission hook, could be a host string, 
+    // a host:port pair, or a URL.
+	Address string
+	// ClientCertificate is the path to a client cert file for TLS.
+	ClientCertificate string
+    // ClientCertificateData contains PEM-encoded data from a client cert file
+    for TLS. Overrides ClientCertificate
+	ClientCertificateData []byte
+	// ClientKey is the path to a client key file for TLS.
+	ClientKey string
+	// ClientKeyData contains PEM-encoded data from a client key file for TLS. Overrides ClientKey
+	ClientKeyData []byte
+	// CertificateAuthority is the path to a cert file for the certificate authority.
+	CertificateAuthority string
+	// CertificateAuthorityData contains PEM-encoded certificate authority certificates. Overrides CertificateAuthority
+	CertificateAuthorityData []byte
+}
+```
+
+## Synchronization of AdmissionControlConfiguration (**optional for alpha implement**)
+
+If the `initializer admission controller` and the `generic webhook admission
+controller` watch the `AdmissionControlConfiguration` and act upon deltas, their
+cached version of the configuration might be arbitrarily delayed. This makes it
+impossible to predicate what initializer/hooks will be applied to newly created
+objects.
+
+We considered a few ways to make the behavior of the `initializer admission
+controller` and the `generic webhook admission controller` predictable.
+
+(I prefer #2. #1 is inefficient, #3 requires complex schema and is not intuitive)
+
+#### 1. Always do consistent read 
+
+The `initializer admission controller` and the `generic webhook admission
+controller` always do consistent read of the `AdmissionControlConfiguration`
+before applying the configuration to the incoming objects. This adds latency to
+every CREATE request. Because the two admission controllers are in the same
+process as the apiserver, the latency mainly consists of the consistent read
+latency of the backend storage (etcd), and the proto unmarshalling.
+
+#### 2. Optimized version of #1, do consistent read of a smaller object
+
+Instead of having the two controllers do consistent read of the entire
+`AdmissionControlConfiguration` object, we let the registry store the
+resourceVersion of the `AdmissionControlConfiguration` (perhaps in a configMap),
```

Comment:
There is no atomic update guarantee right? Apiserver can crash after updating the configmap but before updating AdmissionControlConfiguration.
We should be fine if we always update the configmap first.